train.c

```diff
@@ -8,28 +8,74 @@
 #include <stdlib.h>
 #include <string.h>
 #include <time.h>
-#include "agents/temporal_difference.h"
+
+#include "agents/rl_agent.h"
+#include "agents/util.h"
 #include "game.h"
 
-#define NUM_ITERATION 10000
+// for training
+#define INITIAL_MUTIPLIER 0.0001
+#define LEARNING_RATE 0.02
+#define NUM_EPISODE 10000
+#define EPSILON_GREEDY 0
+#define MONTE_CARLO 1
+
+// for Markov decision model
+#define GAMMA 1
+#define EPISODE_REWARD 1
+
+// for epsilon greedy
 #define EPSILON_START 0.5
 #define EPSILON_END 0.001
+
+#define RAND_UNIFORM ((float) rand() / (float) RAND_MAX)
+
+#if EPSILON_GREEDY
 static double epsilon = EPSILON_START;
 static double decay_factor;
+#endif
+
 static unsigned int N_STATES = 1;
-static td_agent_t agent[2];
-#define RAND_UNIFORM ((float) rand() / (float) RAND_MAX)
+static rl_agent_t agent[2];
+
+_Static_assert(INITIAL_MUTIPLIER >= 0,
+               "Initial mutiplier must not be less than 0");
+_Static_assert(LEARNING_RATE > 0, "Learning rate must be greater than 0");
+_Static_assert(NUM_EPISODE > 0,
+               "The number of episodes must be greater than 0.");
+_Static_assert(EPSILON_GREEDY == 0 || EPSILON_GREEDY == 1,
+               "EPSILON_GREEDY must be a boolean that is 0 or 1");
+_Static_assert(MONTE_CARLO == 0 || MONTE_CARLO == 1,
+               "MONTE_CARLO must be a boolean that is 0 or 1");
+_Static_assert(GAMMA >= 0, "Gamma must not be less than 0");
+_Static_assert(EPISODE_REWARD == 0 || EPISODE_REWARD == 1,
+               "EPISODE_REWARD must be a boolean that is 0 or 1");
+_Static_assert(EPSILON_END >= 0, "EPSILON_END must not be less than 0");
+_Static_assert(EPSILON_START >= EPSILON_END,
+               "EPSILON_START must not be less than EPSILON_END");
+
+
+static void init_agent(rl_agent_t *agent, unsigned int state_num, char player)
+{
+    init_rl_agent(agent, state_num, player);
+    for (unsigned int i = 0; i < state_num; i++)
+        agent->state_value[i] =
+            get_score(hash_to_table(i), player) * INITIAL_MUTIPLIER;
```

Comment:
If we use `check_win` to initialize the values, there may be too many 0s in the early game, resulting in the AI not knowing which move is better. However, if we use `get_score`, the AI always makes the first move at the center, which is usually the best move. This reduces the episode for AI to find the optimal policy. Additionally, we cannot use MCTS because it consumes too much time.